profiling/src/lib.rs

```diff
@@ -0,0 +1,719 @@
+        Tag::new("process_id", unsafe { libc::getpid() }.to_string())
+            .expect("static tags to be valid"),
+    ]
+}
+
+thread_local! {
+    static REQUEST_LOCALS: RefCell<RequestLocals> = RefCell::new(RequestLocals {
+        last_wall_time: Instant::now(),
+        last_cpu_time: None,
+        tags: static_tags(),
+        uri: String::from("http://localhost:8126"),
+    });
+}
+
+/* If Failure is returned the VM will do a C exit; try hard to avoid that,
+ * using it for catastrophic errors only.
+ */
+extern "C" fn rinit(r#type: c_int, module_number: c_int) -> ZendResult {
+    #[cfg(debug_assertions)]
+    trace!("RINIT({}, {})", r#type, module_number);
+
+    unsafe { zend::datadog_php_profiling_rinit() };
+
+    // SAFETY: globals pointer is valid during rinit.
+    let globals = unsafe { zend::datadog_php_profiling_globals_get() };
+    unsafe { read_env(globals) };
+
+    // At the moment, logging is truly global, so init it exactly once whether
+    // profiling is enabled or not.
+    static ONCE: Once = Once::new();
+    ONCE.call_once(|| {
+        #[cfg(not(debug_assertions))]
+        logging::log_init(unsafe { std::mem::transmute(globals.profiling_log_level) });
+        #[cfg(debug_assertions)]
+        log::set_max_level(globals.profiling_log_level);
+
+        unsafe {
+            // reminder: this cannot be done in minit because of Apache forking model
+            PROFILER.replace(Some(Profiler::new()));
+
+            if globals.profiling_enabled {
+                let sapi_module = zend::sapi_module;
+                if sapi_module.pretty_name.is_null() {
+                    let name = CStr::from_ptr(sapi_module.name).to_string_lossy();
+                    warn!("The SAPI module {}'s pretty name was not set!", name)
+                } else {
+                    let pretty_name = CStr::from_ptr(sapi_module.pretty_name).to_string_lossy();
+                    if *SAPI.borrow() != Sapi::Unknown {
+                        debug!("Recognized SAPI: {}.", pretty_name);
+                    } else {
+                        warn!("Unrecognized SAPI: {}.", pretty_name);
+                    }
+                }
+                if let Err(err) = cpu_time::ThreadTime::try_now() {
+                    if globals.profiling_experimental_cpu_time_enabled {
+                        warn!(
+                            "CPU Time collection was enabled but collection failed: {}",
+                            err
+                        );
+                    } else {
+                        debug!(
+                            "CPU Time collection was not enabled and isn't available: {}",
+                            err
+                        );
+                    }
+                } else if globals.profiling_experimental_cpu_time_enabled {
+                    info!("CPU Time profiling enabled.");
+                }
+            }
+        }
+    });
+
+    if globals.profiling_enabled {
+        unsafe {
```

Comment:
This is a _very big_ unsafe block. Can it be made smaller?

Reply:
Yes, I think it can be shrunk quite a bit. In fact, if I switch to using the `once_cell` crate I think all of the unsafe usage of this particular block goes away.